File: Go/113-path-sum-ii.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 var result [][]int

 //dfs方式得到所有路径
 func getPath(root *TreeNode, sum int, res []int) (){
	 res = append(res, root.Val)
	 //只有根结点且满足和
	 if root.Left == nil && root.Right == nil && root.Val == sum {
		 resCopy := make([]int, len(res))
		 copy(resCopy, res)
		 result = append(result, resCopy)
		 return
	 }
	 //递归并回溯
	 sum -= root.Val
	 if root.Left != nil {
		 getPath(root.Left, sum, res)
	 }
	 if root.Right != nil {
		 getPath(root.Right, sum, res)
	 }
 }
 
 func pathSum(root *TreeNode, sum int) [][]int {
	 result = [][]int{}
	 res := []int{}
	 if root == nil {
		 return result
	 }
	 getPath(root, sum, res)
	 return result
 }

File: Python/113-path-sum-ii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    #dfs方式遍历所有路径并判断
    def foundPath(self, root: TreeNode, sum: int, result: List[List[int]], res: List[int]):
        res.append(root.val)
        #只有一个结点且满足和
        if root.left == None and root.right == None and root.val == sum:
            result.append(res.copy())
            return None
        #递归并回溯
        if root.left != None:
            self.foundPath(root.left, sum - root.val, result, res)
            res.pop()
        if root.right != None:
            self.foundPath(root.right, sum - root.val, result, res)
            res.pop()
    def pathSum(self, root: TreeNode, sum: int) -> List[List[int]]:
        result, res = [], []
        if root == None:
            return result
        self.foundPath(root, sum, result, res)
        return result